src/Microsoft.PowerShell.Commands.Utility/commands/utility/PSBreakpointUpdaterCommandBase.cs

```diff
@@ -0,0 +1,168 @@
+
+namespace Microsoft.PowerShell.Commands
+{
+    /// <summary>
+    /// Base class for Enable/Disable/Remove-PSBreakpoint.
+    /// </summary>
+    public abstract class PSBreakpointUpdaterCommandBase : PSBreakpointCommandBase
+    {
+        #region strings
+
+        internal const string BreakpointParameterSetName = "Breakpoint";
+        internal const string IdParameterSetName = "Id";
+
+        #endregion strings
+
+        #region parameters
+
+        /// <summary>
+        /// Gets or sets the breakpoint to enable.
+        /// </summary>
+        [Parameter(ParameterSetName = BreakpointParameterSetName, ValueFromPipeline = true, Position = 0, Mandatory = true)]
+        [ValidateNotNull]
+        public Breakpoint[] Breakpoint { get; set; }
+
+        /// <summary>
+        /// Gets or sets the Id of the breakpoint to enable.
+        /// </summary>
+        [Parameter(ParameterSetName = IdParameterSetName, ValueFromPipelineByPropertyName = true, Position = 0, Mandatory = true)]
+        [ValidateNotNull]
+        public int[] Id { get; set; }
+
+        /// <summary>
+        /// Gets or sets the runspace where the breakpoints will be used.
+        /// </summary>
+        [Parameter(ParameterSetName = IdParameterSetName, ValueFromPipelineByPropertyName = true)]
+        [Alias("RunspaceId")]
+        [Runspace]
+        public override Runspace Runspace { get; set; }
+
+        #endregion parameters
+
+        #region overrides
+
+        /// <summary>
+        /// Gathers the list of breakpoints to process and calls ProcessBreakpoints.
+        /// </summary>
+        protected override void ProcessRecord()
+        {
+            if (ParameterSetName.Equals(BreakpointParameterSetName, StringComparison.OrdinalIgnoreCase))
+            {
+                foreach (Breakpoint breakpoint in Breakpoint)
+                {
+                    if (ShouldProcessInternal(breakpoint.ToString()) &&
+                        TryGetRunspace(breakpoint))
+                    {
+                        ProcessBreakpoint(breakpoint);
+                    }
+                }
+            }
+            else
+            {
+                Debug.Assert(
+                    ParameterSetName.Equals(IdParameterSetName, StringComparison.OrdinalIgnoreCase),
+                    $"There should be no other parameter sets besides '{BreakpointParameterSetName}' and '{IdParameterSetName}'.");
+
+                foreach (int id in Id)
+                {
+                    Breakpoint breakpoint;
+                    if (TryGetBreakpoint(id, out breakpoint) &&
+                        ShouldProcessInternal(breakpoint.ToString()))
+                    {
+                        ProcessBreakpoint(breakpoint);
+                    }
+                }
+            }
+        }
+
+        #endregion overrides
+
+        #region private data
+
+        private readonly Dictionary<Guid, Runspace> runspaces = new Dictionary<Guid, Runspace>();
```

Comment:
If runspace references are only kept for the command instance then that should be Ok.  But I wonder if looking up from this dictionary is noticeably faster than looking up from the RunspaceDictionary.

Reply:
That's a good question. Maybe the optimization isn't worth it, and maybe it is already being handled well enough via the RunspaceDictionary as you suggest. If that's the case, I can just drop the use of the private dictionary.

Reply:
I just realized this PR was marked as waiting on author, so I guess y'all are waiting for me to respond further here. I reviewed the code in question, and here is what I found:

1. With the current approach, the cmdlet creates a dictionary and caches runspaces by their GUID for faster retrieval when multiple breakpoints from a single runspace are piped into a command. The cache is only for the cmdlet instance and therefore references will not be retained beyond the invocation of the cmdlet.
1. If I remove the caching, then every lookup of the runspace by ID requires locking a sync object, walking through a sorted dictionary of runspaces, adding them to a list, returning a read-only collection that is built using that list, and then walking through that collection until one is found that has the same GUID instance id, plus a few additional things like creating an array to pass in the instance id.

I haven't specifically tested performance between those two, but it seems clear to me that the former is a worthwhile optimization to have given the work that is done in the latter, especially when the code that is added is so trivial (a simple dictionary with lookup/caching performed inside of a single method).

Do we need me to perform performance testing here (it hardly seems worth the time), or can this portion be marked resolved and the "waiting on author" tag removed so that this PR can move forward?